### DISCUSSION TITLE
k3s   How should I replace iptables with ipvs

### DISCUSSION THREAD

**Discussion by MagicStarTrace** [k3s-io/k3s#4412] [Q&A]
I installed k3s  , The environment and installation commands are as follows

os   version CentOS Linux
k3s version v1.21.5+k3s2
docker version: 20.10.10

curl -sfL http://rancher-mirror.cnrancher.com/k3s/k3s-install.sh | INSTALL_K3S_MIRROR=cn sh -s - --docker  --write-kubeconfig ~/.kube/config 

I now plan to use ipvs to replace iptables, but since kube-proxy is not installed, how to replace it?

Thank You！

**Comment by brandond** [comment]
kube-proxy is built in to k3s. You can pass extra args to it via: `--kube-proxy-arg=proxy-mode=ipvs`. This is documented at:
https://rancher.com/docs/k3s/latest/en/installation/install-options/agent-config/#customized-flags

**Reply by MagicStarTrace** [reply]
Use this command ? 

curl -sfL  http://rancher-mirror.cnrancher.com/k3s/k3s-install.sh|INSTALL_K3S_EXEC="server --docker" INSTALL_K3S_MIRROR=cn  sh -s -  --kube-proxy-arg=proxy-mode=ipvs --disable=metrics-server --disable-selinux   --write-kubeconfig  ~/.kube/config 

![image](https://user-images.githubusercontent.com/15685376/140615757-d53691a3-a77c-4d77-8551-d109dd464684.png)

I can still see that there are iptables rules. Is the method wrong or is it impossible to completely disable iptables?

In addition, how is the agent configured to use ipvs by default , Use this command ? 
curl -sfL  http://rancher-mirror.cnrancher.com/k3s/k3s-install.sh|INSTALL_K3S_MIRROR=cn K3S_URL=https://192.168.100.20:6443 K3S_TOKEN=K10aa588e3eb469fa84ce375bb2ca993ccdaf97713cb2c7b447da9ef95b6d145f79::server:afcaf7b294fb551f6a5caa7fd49b21f7  sh -s - --docker  --kube-proxy-arg=proxy-mode=ipvs

Thank You!

**Comment by brandond** [comment]
>I can still see that there are iptables rules. Is the method wrong or is it impossible to completely disable iptables?

Changing the mode to ipvs does not eliminate ALL use of iptables, it just uses ipvs for kube-proxy's routing to Kubernetes services instead of iptables. All the other stuff that uses iptables (kubelet, flannel, kube-router, etc) is still there.

**Reply by MagicStarTrace** [reply]
So, am I successful in modifying it to ipvs? 
Thank you!
